cmd/package_managers.go

```diff
@@ -138,3 +176,60 @@ func fetchGitRepositoryFromRubyGems(packageName string, manager packageManagerCl
 	}
 	return v.SourceCodeURI, nil
 }
+
+// Gets the GitHub repository URL for the nuget package.
+func fetchGitRepositoryFromNuget(packageName string, manager packageManagerClient) (string, error) {
+	nugetIndexURL := "https://api.nuget.org/v3/index.json"
+	respIndex, err := manager.GetURI(nugetIndexURL)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("failed to get nuget index json: %v", err))
+	}
+	defer respIndex.Body.Close()
+	nugetIndexResults := &nugetIndexResults{}
+	err = json.NewDecoder(respIndex.Body).Decode(nugetIndexResults)
+
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("failed to parse nuget index json: %v", err))
+	}
+	packageBaseAddressIndex := slices.IndexFunc(nugetIndexResults.Resources,
+		func(n nugetIndexResult) bool { return n.Type == "PackageBaseAddress/3.0.0" })
+	if packageBaseAddressIndex == -1 {
+		return "", sce.WithMessage(sce.ErrScorecardInternal, "failed to find package base URI at nuget index json")
+	}
+
+	nugetPackageBaseURL := nugetIndexResults.Resources[packageBaseAddressIndex].ID
+
+	respPackageIndex, err := manager.Get(nugetPackageBaseURL+"%s/index.json", packageName)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("failed to get nuget package index json: %v", err))
+	}
+	defer respPackageIndex.Body.Close()
+	packageIndexResults := &nugetpackageIndexResults{}
+	err = json.NewDecoder(respPackageIndex.Body).Decode(packageIndexResults)
+
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("failed to parse nuget package index json: %v", err))
+	}
+
+	lastVersion := packageIndexResults.Versions[len(packageIndexResults.Versions)-1]
```

Comment:
The ordering of versions in the index.json document here is not well defined (note there is not reference ordering in the protocol docs: https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#enumerate-package-versions). NuGet.org may put them in SemVer order today (and will perhaps always do this to avoid unnecessary breaking changes) but it's not something that should be depended on if there's a reasonable alternative. Also, the latest (or any version in this document) version may be unlisted. In NuGet. the standard definition of "latest" is "highest SemVer version that is not unlisted from search and is available (not deleted)". 

I don't know there full intent here, but it may be necessary to go to the package metadata endpoint so you can filter out latest unlisted versions.

Reply:
@joelverhagen thanks for the review and correcting the ordering of versions proposed at the first iteration.
please review the new implementation and the tests that show how this works with the nuget protocol.
leaving this un-resolved until your review.